daemon/SecretsImpl/dataprotector.cpp

```diff
@@ -0,0 +1,326 @@
+        QString path = fileInfo.absoluteFilePath();
+        QSharedPointer<QFile> file(new QFile(path));
+        file->open(QIODevice::ReadOnly);
+        return file;
+    });
+
+    // Open all files
+    bool noFilesOpen = std::all_of(files.begin(), files.end(), [](QSharedPointer<QFile> &file) {
+        bool isOpen = file->isOpen();
+        if (!isOpen) {
+            // Loss of any individual file here doesn't yet necessarily mean an error
+            qWarning(lcSailfishSecretsDaemon) << "can't open file:" << file->fileName();
+        }
+        return !isOpen;
+    });
+
+    // If none of the files could be opened, we are very unlucky
+    if (noFilesOpen) {
+        qWarning(lcSailfishSecretsDaemon) << "Could not open any files. Data is irretrieveable.";
+        return Irretrievable;
+    }
+
+    // Read all files
+    QVector<QByteArray> fileContents;
+    std::transform(files.begin(), files.end(), std::back_inserter(fileContents), [](QSharedPointer<QFile> &file) {
+        return file->readAll();
+    });
+
+    std::remove_if(fileContents.begin(), fileContents.end(), [](QByteArray &byteArray) {
+        return byteArray.isEmpty();
+    });
+
+    //  If none of the files could be read, again we are unlucky
+    if (fileContents.size() == 0) {
+        qWarning(lcSailfishSecretsDaemon) << "Could not read any of the files. Data is irretrievable.";
+        return Irretrievable;
+    }
+
+    // Assuming that any of the byte arrays can differ from any of the others,
+    // count how many times each occours. Ideally we end up with 3 occourences
+    // on the first index, and 0 on the others.
+    QVector<int> occourences(fileContents.size());
+    for (const QByteArray &currentData : fileContents) {
+        for (int i = 0; i < fileContents.size(); i++) {
+            if (currentData == fileContents.at(i)) {
+                occourences[i]++;
+                break;
+            }
+        }
+    }
+
+    // Assume that the version that occours the most is the correct version.
+    int maxOccourenceIndex = 0;
+    for (int i = 0; i < occourences.size(); i++) {
+        if (occourences[i] > occourences[maxOccourenceIndex]) {
+            maxOccourenceIndex = i;
+        }
+    }
+    int minimumValidOccourences = fileContents.size() / 2 + 1;
+    if (occourences[maxOccourenceIndex] < minimumValidOccourences) {
+        qWarning(lcSailfishSecretsDaemon) << "Could not decide which file is correct, there is no majority.";
+        return Irretrievable;
+    }
+
+    m_data = fileContents[maxOccourenceIndex];
+    *result = m_data;
+
+    return Success;
+}
+
+DataProtector::Status DataProtector::putData(const QByteArray &bytes)
+{
+    QDir dir(m_path);
+
+    // Check if the main directory exists, create if not
+    if (!dir.exists()) {
+        if (!dir.mkpath(m_path)) {
+            qCWarning(lcSailfishSecretsDaemon) << "Can't create protected root directory:" << m_path;
```

Comment:
Use different wording here so that getData and putData are not logging the same error message e.g. "Can't create protected root directory when writing:" << m_path;

Reply:
Will fix.